### PR TITLE
feat(network): O-4b — network join consumes a leaf package (--from-package)

### LINKED ISSUE #1063: O-4b: cortex network join consumes the leaf package from the provisioning response
Per design #1049 §3 + O-4 (#1054) decomposition. Make `cortex network join` source the leaf package (operator JWT + account + account JWT + creds path) from a provisioning response / descriptor field (or a `--from-package <file>` for the interim) and feed it to O-3's (#1058) `OperatorModeLeafPackage` conversion seam — so the operator doesn't pass the four flags by hand. cortex-local, builds on the O-3 branch's seam, fully testable (no production touch). Branch off O-3's branch (feat/o3-join-autoconvert) to consume the package type. Part of EPIC #1050; sibling O-4a (#TBD) supplies the package over the wire.

### PR DESCRIPTION
**Pull request by mellanon**
## What

`cortex network join` now **sources the operator-mode leaf package from a file** (`--from-package <file>`) instead of demanding the four `--operator-jwt` / `--account-jwt` / `--account` / `--system-account*` flags by hand. The package is the interim, on-disk form of the **exact shape O-4a's signed `register→issue` response will carry** — so this is the **consumer half** of O-4's handshake (`#1050` spec §3, D3), landing ahead of the wire transport.

## Wire shape — the single O-4a-produces / O-4b-consumes contract

```ts
interface LeafPackageFile {
  operatorJwt: string;       // NSC operator JWT (eyJ…) — root of the account tree
  account: string;           // issued account pubkey (nkey-U, A…) the leaf binds to
  accountJwt: string;        // issued account JWT (eyJ…) — preloaded under resolver_preload
  systemAccount?: string;    // OPTIONAL SYS account pubkey (nkey-U) → system_account
  systemAccountJwt?: string; // OPTIONAL SYS account JWT (preloaded alongside)
  credsPath: string;         // REQUIRED issued leaf .creds PATH (the file is the secret, not the path)
  endpoint?: string;         // OPTIONAL advisory dial-URL hint; the authoritative hub_url is the SIGNED descriptor's (DD-12)
}
```

Superset of O-3's `OperatorModeLeafPackage` (the render material) + `credsPath` (the issued creds — the point of the handshake) + advisory `endpoint`. All fields are **public** key material (JWTs + nkey-U pubkeys); never a seed.

## Source / precedence / validation

- **Source:** `network-leaf-package.ts` — `parseLeafPackageFile(json)` (pure: string in, result out). `network.ts` reads + parses the file **first**, as a pure read.
- **Precedence:** `explicit flags > --from-package file > config > convention`, per field (`operatorJwt` / `account` / `accountJwt` / `systemAccount*` / `credsPath`). The package sets only what it carries; a flag always wins; the package beats config. Threaded as a `JoinOverrides` layer in `deriveJoinInputs`.
- **Validation — REUSE, not duplicate:** the nkey-U / JWT-shape guards are O-3's, now exported from `leaf-remote-renderer` as `isNkeyAccountPubkey` / `isNscJwtShape` and called by the parser (one grammar, no drift). A malformed package (bad nkey/JWT, missing required field, half-specified SYS account, non-object JSON) **fails fast as a usage error BEFORE any derivation or mutation** — unvalidated material never reaches the operator-mode conversion seam. **Dry-run stays a pure read.**

## Tests (+26)

- **Parser (17):** minimal/full happy path, forward-compat (unknown fields ignored), projection (`leafPackageToOperatorMode` drops `credsPath`/`endpoint`), and fail-fast on every malformed shape.
- **Derive (6):** package fields flow into the conversion; `credsPath` flows; SYS carried; **flags override package**; **package overrides config**; none → undefined.
- **CLI (3):** malformed package → exit 2, **no write**; missing file → exit 2; **dry-run with a valid package writes nothing** (pure read).

All fixtures are obvious fake `eyJ…` / `A…` material — **zero seeds, zero real keys**.

## Gates

- `tsc --noEmit` clean
- `eslint --quiet src/` clean
- `bun test src/` — **6886 pass / 0 fail / 2 skip**
- `check-carveouts.sh` — PASS

## Stacking

**Stacked on O-3 (#1058)** — base is `feat/o3-join-autoconvert`, NOT `main`. O-4b consumes O-3's `OperatorModeLeafPackage` type + the renderer/seam + the atomic-write fix. Merge O-3 first.

Closes #1063
Refs #1050, #1054, #1049

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/cli/cortex/commands/network-leaf-package.ts [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+  }
+  if (!isNscJwtShape(accountJwt)) {
+    return {
+      ok: false,
+      reason: "leaf package `accountJwt` is not a JWT (expected an `eyJ…` token with three dot-separated segments)",
+    };
+  }
+
+  // 5) Required creds path. The PATH (not the creds) — must be a non-empty string.
+  const credsPath = optionalString(obj.credsPath);
```
Comment (mellanon) [inline src/cli/cortex/commands/network-leaf-package.ts:193]: **[nit/security]** `credsPath` is the only required string field that does NOT call `.trim()` after `optionalString()`. Lines 157, 169, 181, 204, 205 all call `?.trim()`; line 193 does not.

A `credsPath` value with leading/trailing whitespace (e.g. `"  ~/.config/nats/issued.creds  "`) passes the parser as valid, then fails at `existsSync` (path not found) or at the renderer's `startsWith("/")` guard — but with a less actionable error than a parser-level rejection.

**Fix:** apply `.trim()` consistently — `const credsPath = optionalString(obj.credsPath)?.trim()` — and add a test case for a whitespace-padded `credsPath`.

File: src/cli/cortex/commands/network-leaf-package.ts [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+// =============================================================================
+// Parse + validate.
+// =============================================================================
+
+/** Read a non-empty string field; returns `undefined` when absent/empty/wrong-type. */
+function optionalString(value: unknown): string | undefined {
+  if (typeof value !== "string") return undefined;
+  const trimmed = value.trim();
+  return trimmed.length === 0 ? undefined : value;
```
Comment (mellanon) [inline src/cli/cortex/commands/network-leaf-package.ts]: **[nit/quality]** `optionalString()` returns the UNTRIMMED original value when the string is non-empty (line 112: `return trimmed.length === 0 ? undefined : value`). Callers immediately call `?.trim()` to compensate. This is correct for the grammar-validated fields but creates the `credsPath` inconsistency noted above.

Consider simplifying by making `optionalString` return the trimmed value directly (`return trimmed.length === 0 ? undefined : trimmed`), eliminating the double-trim and making the contract obvious. Each callers' `?.trim()` would then be a no-op and could be removed. Low priority — existing behavior is correct for all validated fields.

File: src/cli/cortex/commands/network-leaf-package.ts [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+  };
+}
+
+/**
+ * Project a validated {@link LeafPackageFile} onto the {@link
+ * OperatorModeLeafPackage} O-3's renderer consumes. The package is a SUPERSET
+ * (it also carries `credsPath` + `endpoint`, which are NOT operator-mode render
+ * material), so this drops those two and returns only the render fields.
+ */
+export function leafPackageToOperatorMode(
```
Comment (mellanon) [inline src/cli/cortex/commands/network-leaf-package.ts:256]: **[nit/security]** `leafPackageToOperatorMode` is exported but only appears in tests — not in the production join path. The production path correctly uses field-by-field spreading (`pkg?.operatorJwt`, etc.) in `deriveJoinInputs`. The exported helper is fine for future callers (O-4a will use it), but it's worth a doc comment clarifying: "not yet used in the production path; projected for O-4a wire consumption". Low priority.

**Review by mellanon — COMMENTED** [review]
Lenses applied: CodeQuality, Security (targeted). Gates: tsc clean, eslint clean, 82 targeted tests pass (17 parser + 35 derive + 30 CLI), carve-outs pass.

## Adversarial security points (#1–5)

**#1 — Validation before mutation (PASS)**
The package read is a pure `readFileSync` → `parseLeafPackageFile` gate in `readLeafPackageFlag` (network.ts:337–361), called BEFORE `deriveJoinInputs`, before any derivation, before any file write. `parseLeafPackageFile` validates every JWT field against `isNscJwtShape` (the same `JWT_SHAPE = /^eyJ[A-Za-z0-9_-]+(?:\.[A-Za-z0-9_-]+){2}$/` the renderer uses) and every nkey-U against `isNkeyAccountPubkey` (same `NKEY_ACCOUNT = /^A[A-Z2-7]{55}$/`). These are not local copies — they are the exported predicates from `leaf-remote-renderer.ts` lines 962–973, so grammar drift is structurally impossible. A malformed package (bad JWT, bad nkey, missing required field, half-specified SYS account, non-object JSON) returns `{ ok: false }` → `usageError` (exit 2) before `deriveJoinInputs` is called. JWTs emitted bare into HOCON are bounded to base64url+dot, no whitespace/braces; account nkeys are bounded to `A[A-Z2-7]{55}`, no injection surface. `credentials` in the HOCON output goes through `JSON.stringify` so a newline-containing `credsPath` cannot break out. No unvalidated package field reaches the conversion seam.

**#2 — Precedence (PASS)**
`flag > package > config > convention` is correct field-by-field in `deriveJoinInputs` (network-derive.ts:388–410). An absent package field does not clobber a flag/config value: `overrides.operatorJwt ?? pkg?.operatorJwt ?? natsInfra?.operator_jwt`. Confirmed by 4 derive tests (flag wins, package wins over config, package credsPath flows, none → undefined).

**#3 — Wire-shape contract (PASS)**
`LeafPackageFile` is the correct O-4a→O-4b contract: superset of `OperatorModeLeafPackage` + `credsPath` (REQUIRED, the whole point of the handshake) + `endpoint` (advisory, NOT bound into the descriptor, per DD-12 — hub_url comes from the signed network descriptor). `leafPackageToOperatorMode` correctly drops both non-render fields.

**#4 — Dry-run writes nothing (PASS)**
`readLeafPackageFlag` is a synchronous pure read (`readFileSync`). When `apply=false` (the default), `buildDryRunPorts` is used — no writes. CLI test `O-4b dry-run with a VALID package writes NOTHING (pure read)` verifies this end-to-end.

**#5 — No key material in fixtures (PASS)**
Zero `S[A-Z2-7]{56}` seed patterns found in any of the 3 new test files. All fixtures use obvious fakes: `eyJhbGci…` JWTs and `A` + repeated base32 chars.

## Findings

**NIT-1** (network-leaf-package.ts:193) — `credsPath` is the only required field that does NOT call `.trim()` after `optionalString()`. All other fields (lines 157, 169, 181, 204, 205) call `?.trim()`. A whitespace-padded path passes the parser but fails later at `existsSync` with a less actionable error. Fix: `const credsPath = optionalString(obj.credsPath)?.trim()`. Posted as inline comment.

**NIT-2** (network-leaf-package.ts:112) — `optionalString` returns the UNTRIMMED `value` (not `trimmed`), so callers all call `.trim()` again. Simplify by returning `trimmed` directly. Low priority, no behavior impact for validated fields. Posted as inline comment.

**NIT-3** (network-leaf-package.ts:256) — `leafPackageToOperatorMode` is exported but only used in tests today. A doc comment noting it is projected for O-4a consumption would help future readers. Author's call.

verdict: blockers=0 majors=0 nits=3 — recommend: request-changes. One nit is worth a one-line fix (credsPath trim) before merge; the other two are cosmetic. The security architecture is sound. (posted as comment-review — bot account opened the PR; --request-changes blocked by GitHub)

**Comment by mellanon** [comment]
**Review complete** — 3 inline comments + verdict submitted above. Discord: Echo reviewed O-4b; 0 blockers, 0 majors, 3 nits (NIT-1 credsPath trim is the one worth fixing before merge). Security architecture is sound.

**Comment by mellanon** [comment]
Review nits addressed (0 blockers/majors — security architecture confirmed sound):

| Finding | Outcome |
|---|---|
| NIT-1 credsPath not trimmed | **Fixed** — root cause was optionalString returning the untrimmed value; now returns trimmed, so credsPath (+ all fields) trim uniformly |
| NIT-2 optionalString returns untrimmed | **Fixed** — returns `trimmed`; the 5 now-redundant per-caller `.trim()` calls dropped |
| NIT-3 leafPackageToOperatorMode O-4a doc | **Covered** — module header documents the O-4a contract + the projection's role |

Behavior-preserving. Gates (stacked PR, no CI — local): tsc ✓ · eslint ✓ · bun test src/ ✓ · check-carveouts ✓. Signed.

**Comment by mellanon** [comment]
Superseded by #1090 (this PR auto-closed when its stacked base `feat/o3-join-autoconvert` was deleted on #1058's merge; GitHub won't reopen a PR with a deleted base). #1090 carries the identical work retargeted to `main`.
